File: Python/src/Day2.py
import os.path
import Util
#varialbles
debug = False #set to true for debug messages

#File Directories
root = os.path.join(os.path.abspath(os.getcwd()),"Python")
resc = os.path.join(root,"Ressources")

#Functions
def isSafe(lineOfCode):
    increasing = True
    safe = True

    if(lineOfCode[0]>lineOfCode[1]):
        increasing = False
    if debug:
        print(f"The Code of {lineOfCode} is {"increasing" if increasing else "decreasing"}")
    for i in range(len(lineOfCode)-1):
        if(safe and increasing):
            if(lineOfCode[i+1]<=lineOfCode[i]):
                safe = False
                if debug:
                    print(f"The Code is unsafe because the {i+1}{"st."if i==0 else "nd." if i==1 else "rd." if i==2 else "th."} element of {lineOfCode[i+1]} is not higher than the {i}{"st."if i==0 else "nd." if i==1 else "rd." if i==2 else "th."} element of {lineOfCode[i]}")
            elif(lineOfCode[i+1]>lineOfCode[i]+3):
                safe = False
                if debug:
                    print(f"The Code is unsafe because the {i+1}{"st."if i==0 else "nd." if i==1 else "rd." if i==2 else "th."} element of {lineOfCode[i+1]} increases by more than 3 upon the {i}{"st."if i==0 else "nd." if i==1 else "rd." if i==2 else "th."} element of {lineOfCode[i]}")

        if(safe and not increasing):
            if(lineOfCode[i+1]>=lineOfCode[i]):
                safe = False
                if debug:
                    print(f"The Code is unsafe because the {i+1}{"st."if i==0 else "nd." if i==1 else "rd." if i==2 else "th."} element of {lineOfCode[i]} is not lower than the {i}{"st."if i==0 else "nd." if i==1 else "rd." if i==2 else "th."} element of {lineOfCode[i-1]}")
            elif(lineOfCode[i+1]<lineOfCode[i]-3):
                safe = False
                if debug:
                    print(f"The Code is unsafe because the {i+1}{"st."if i==0 else "nd." if i==1 else "rd." if i==2 else "th."} element of {lineOfCode[i]} dencreases by more than 3 upon the {i}{"st."if i==0 else "nd." if i==1 else "rd." if i==2 else "th."} element of {lineOfCode[i-1]}")
    if (safe and debug):
        print(f"The code of {lineOfCode} is safe")
    return safe
        
def isSafeWithDampener(lineOfCode):
    safe = False #Assuming it is unsafe. Should we find a safe variant it will be overridden
    for i in range(len(lineOfCode)):
        if(not safe): #Stop when safe variant was found
            adjustedCode = []
            if debug:
                print(f"Checking if the code of {lineOfCode} would be safe without the {i+1}{"st."if i==0 else "nd." if i==1 else "rd." if i==2 else "th."} digit of {lineOfCode[i]}")
            for j in range(len(lineOfCode)):
                #Building new line of Code:
                if(i>j):
                    adjustedCode.append(lineOfCode[j])
                elif(i<j):
                    adjustedCode.append(lineOfCode[j])
            if(isSafe(adjustedCode)):
                safe = True
                if debug:
                    print(f"Safe Variant was found")
    return safe

print("Advent of Code day 2:")
inputFile="input2.txt"

code = Util.parseInt(Util.splitLines(Util.importText(os.path.join(resc,inputFile))))

if(debug):
    for line in code:
        print(line)

safeCount = 0
for line in code:
    if(isSafe(line)):
        safeCount=safeCount+1
        if debug:
            print(f"A safe line was found")
    else:
        if(isSafeWithDampener(line)):
            safeCount = safeCount+1
            if debug:
                print(f"Using the Problem Dampener, a Safe line was found")

print(f"There are {safeCount} safe lines of code to be found.")

File: Java/src/Day2.java
import helper.ImportUtility;
import helper.StringUtility;

public class Day2 
{
    static final boolean DEBUG = false;

    public static void run(String fileName)
    {
        String input[] = ImportUtility.getLines(Main.RessourcePath+fileName);
        if(input!=null)
        {
            String lines[][] = ImportUtility.splitColums(input);
            int numbers[][] = new int[lines.length][];

            for(int i = 0; i < lines.length; i++)
            {
                numbers[i] = StringUtility.convertToInt(lines[i]);
            }

            int count = 0;
            for(int[] row : numbers)
            {
                boolean safe = isSafe(row);

                if(safe)
                    count++;
                else
                {
                    safe = isSafeWithDampener(row);
                    if(safe)
                    {
                        count++;
                    }
                }
                
                if(DEBUG)
                    System.out.printf("The line is " + (safe?"safe":"unsafe") + "\n");
            }
            System.out.printf("%n there were %d safe lines.%n", count);
        }
    }
    

    private static boolean isSafe(int[] arrayIn)
    {
        boolean increases = true;
        boolean safe = true;
        if(arrayIn.length >1)
        {
            if(DEBUG)
            {
                System.out.printf("Checking if the series of");
                for(int i : arrayIn)
                    System.out.printf(" %d,", i);
                System.out.printf(" is Safe.%n");
            }
            //check if array increases or decreases. In case of no difference, it will be declared unsafe anyways
            if(arrayIn[0] > arrayIn[1])
                increases = false;
            if(DEBUG)
                System.out.printf("The line " + (increases?"increases%n":"decreases%n"));

            for(int i = 1; i < arrayIn.length; i++)
            {
                if(increases)
                {
                    if(arrayIn[i] <= arrayIn[i-1])
                    {
                        safe = false;
                        if(DEBUG)
                            System.out.printf("unsafe because %d is not higher than %d%n", arrayIn[i], arrayIn[i-1]);
                    }
                    if(arrayIn[i] > arrayIn[i-1]+3)
                    {
                        safe = false;
                        if(DEBUG)
                            System.out.printf("unsafe because %d is more than 3 higher than %d%n", arrayIn[i], arrayIn[i-1]);
                    }
                }
                else if(!increases)
                {
                    if(arrayIn[i] >= arrayIn[i-1])
                    {
                        safe = false;
                        System.out.printf("unsafe because %d is not lower than %d%n", arrayIn[i], arrayIn[i-1]);
                    }
                    if(arrayIn[i] < arrayIn[i-1]-3)
                    {
                        safe = false;
                        System.out.printf("unsafe because %d is more than 3 lower than %d%n", arrayIn[i], arrayIn[i-1]);
                    }
                }
            }
        }
        return safe;
    }

    private static boolean isSafeWithDampener(int[] arrayIn)
    {
        if(DEBUG)
        {
            System.out.printf("Check Problem Dampener for");
            for(int part : arrayIn)
                System.out.printf(" %d,",part);
            System.out.printf("%n");
        }
        boolean safe = false;
        for(int i=0; i<arrayIn.length; i++)
        {
            if(DEBUG)
                System.out.printf("Check if line is safe without %d%n", arrayIn[i]);
            int[] arrayNew = new int[arrayIn.length-1];
            
            for(int j=0; j<arrayIn.length; j++)
            {
                if(i!=j)
                {
                    if(j<i)
                        arrayNew[j]=arrayIn[j];
                    else if(j>i)
                        arrayNew[j-1]=arrayIn[j];
                }
                else if(DEBUG) System.out.printf("Skipping the %d" + (j==0 ? "st" : j==1 ? "nd" : j==2 ? "rd" : "th") + " element of %d%n", j+1, arrayIn[j]);
            }

            if(isSafe(arrayNew))
            {
                safe = true;
                if(DEBUG)
                {
                    System.out.printf("The Line of");
                    for(int part : arrayIn)
                        System.out.printf(" %d,", part);
                    System.out.printf(" without %d is safe!%n", arrayIn[i]);
                }
            }
        }
        return safe;
    }
}
